Add database

Iogamy opened the pull request.

Iogamy committed:
Filmorate Sprint11 1st iteration

Iogamy committed:
Filmorate Sprint11 1st iteration

Ivanov-s-tmn commented on diagram.png:
Отличная структура БД!

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -4,57 +4,62 @@
-@Slf4j
 @RestController
-@RequiredArgsConstructor
+@Slf4j
```
Хорошо покрыл логами эндпоинты!

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/ValidationExceptionHandler.java:
Добавь обработчик для NotFoundException и возвращай 400 код. И общий обработчик для INTERNAL_SERVER_ERROR (500 код), лови в нем все исключения `Throwable e`

Iogamy replied:
ResponseEntity.class отлавливает же эти исключения). Или я не прав?)

Ivanov-s-tmn replied:
У тебя написан один метод-обработчик, который ловит только `ValidationException`, можно написать такие же для `NotFoundException`, потому что ты выбрасываешь его в своем коде и для отлова общих исключений, но это не обязательно)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/mapper/FilmMapper.java:
Поправь выравнивание кода в классе

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/GenreStorage.java:
```diff
@@ -0,0 +1,76 @@
+@RequiredArgsConstructor
+public class GenreStorage {
+    private final JdbcTemplate jdbcTemplate;
+    GenreMapper genreMapper = new GenreMapper();
```
Через внедрение зависимостей не работает, как с остальными классами? Не забывай модификаторы

Iogamy replied:
Не совсем понял вопрос)

Ivanov-s-tmn replied:
`GenreMapper genreMapper = new GenreMapper();` Просто ты явно создаешь объект через new, как до спринга было)
Со спрингом можно передать управление за созданием и внедрением зависимостей спрингу, т.е. ты просто задаешь конструктор или аннотацию `@Autowired` для спринга и он уже самостоятельно создает и внедряет зависимость)
В  твоем случае используется `@RequiredArgsConstructor`, эта аннотация создает конструктор для инициализации final полей, тебе просто нужно сделать поле `private final GenreMapper genreMapper;` и созданием займется спринг)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java:
```diff
@@ -0,0 +1,49 @@
+    private final LikeStorage likeStorage;
+
+
+    public void addLike(Long filmId, Long userId) {
+        likeStorage.addLike(filmId, userId);
+    }
+
+    public void deleteLike(Long filmId, Long userId) {
+        likeStorage.deleteLike(filmId, userId);
+    }
+
+    public List<Film> getPopular(Integer count) {
+        return likeStorage.getPopular(count);
+    }
+
+    public List<Film> getFilms() {
+        return filmStorage.getFilms();
+    }
+
+    public Film create(Film film) {
+        return filmStorage.create(film);
+    }
+
+    public Film update(Film film) {
+        return filmStorage.update(film);
+    }
+
+    public Film getFilmById(Long id) {
+        return filmStorage.getFilmById(id);
+    }
+
+    public Film delete(Long id) {
+        return filmStorage.delete(id);
+    }
```
Вынеси основную логику в сервисы, например получение пользователя и фильма из репозитория перед вызовом метода добавления/удаления лайка, в остальных сервисах аналогично

Iogamy replied:
Запутался что-то ))) Можно пример?)

Ivanov-s-tmn replied:
```
public void addLike(Long filmId, Long userId) {
        final Film film = filmStorage.get(filmId);
	final User user = userStorage.get(userId);
        likeStorage.addLike(film.getId(), user.getId());
}
```
Такая реализация логики в сервисном слое позволит тебе избавиться от зависимостей в сторейдж классах от других сторейдж классов и обращаться в сервисах к сторейдж слою. В других методах принцип аналогичный, т.е. ты обращаешься к сторейдж классу за нужными данными из БД, а дальше по необходимости описываешь логику и снова обращаешься к сторейдж слою для записи в БД, принцип примерно такой)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -0,0 +1,114 @@
+        film.setDescription(filmColumn.getDescription());
+        film.setDuration(filmColumn.getDuration());
+        film.setReleaseDate(filmColumn.getReleaseDate());
+        film.setMpa(mpaStorage.getMpaById(filmColumn.getMpaId()));
```
Смотри, тут будет подзапрос при создании фильма. Ты же можешь сразу достать фильм с его рейтингом, между таблицами есть связь по mpa_id, сделай join и получишь фильм с рейтингом)

Iogamy replied:
Тут не совсем понял)

Ivanov-s-tmn replied:
Смотри, пример запроса `SELECT * FROM films f, mpa_ratings m where f.rating_id = m.id and f.id = ?` Вместе с фильмом подтянется и его рейтинг из таблицы рейтингов благодаря внешнего ключу rating_id и не нужно делать отдельный подзапрос. В этом запросе происходит cross join, когда таблицы указываются через запятую, можно сделать явный join:
`SELECT * FROM films f JOIN mpa_ratings m ON f.rating_id = m.id WHERE f.id = ?`

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/interfaces/FilmStorage.java:
```diff
@@ -0,0 +1,19 @@
+
+    List<Film> getFilms();
+
+    Film create(@Valid Film film);
+
+    Film update(@Valid Film film);
```
валидация в storage слое не нужна

Iogamy replied:
Убрал валидацию) в юзерах - аналогично)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/user/InMemoryUserStorage.java:
InMemory сторейджи можно удалить) Вместе с тестами

Iogamy committed:
Filmorate Sprint11 2nd iteration
Поправил замечания, за исключением некоторых моментов, так как боюсь, что может посыпаться рабочая логика)
